Added minimum edit/Levenshtein distance based alignment function with doctest

okz12 opened the pull request:
Created this for feature/issue #2221, a minimum edit-distance based alignment of two strings. The implementation reuses edit_distance functions in nltk.metrics but adds backtrace to be able to recover the original strings with minimum distance mappings.

E.g., 'f', 'bar' will align in 'fubar' and 'foobar'. 

`print(edit_distance_align('fubar', 'foobar'))`
> [(0, 0), (1, 1), (1, 2), (2, 3), (3, 4), (4, 5), (5, 6)]
NB: 0 is a null/stop index

The mapping from first string to second printed out:
```
s1 = '_fubar'
s2 = '_foobar'
print([(s1[x[0]], s2[x[1]]) for x in edit_distance_align('fubar','foobar')])
```
> [('_', '_'), ('f', 'f'), ('f', 'o'), ('u', 'o'), ('b', 'b'), ('a', 'a'), ('r', 'r')]

okz12 committed:
Added edit distance based alignment function with doctest

iliakur commented on nltk/metrics/distance.py:
```diff
@@ -103,6 +103,76 @@ def edit_distance(s1, s2, substitution_cost=1, transpositions=False):
+
+    while (i, j) != (0, 0):
+        # find direction of minimum cost
+        s1_skip_cost = s2_skip_cost = sub_cost = float('inf')
```
May I suggest a more declarative take on the inside of the while-loop? It takes advantage of Python's built-in `min` function and the `operator` module.

```python
directions = [
    (i - 1, j),      # skip s1
    (i, j - 1),      # skip s2
    (i - 1, j - 1),  # substitution
]
direction_costs = (
   (lev_map[i][j], (i, j)) for i, j in directions
)
_, (i, j) = min(direction_costs, key=operator.itemgetter(0))
```

Feel free to ask if anything in the snippet above is unclear!

okz12 replied:
I've made this change.

iliakur commented on nltk/metrics/distance.py:
```diff
@@ -103,6 +103,76 @@ def edit_distance(s1, s2, substitution_cost=1, transpositions=False):
+            i, j = i - 1, j
+
+        alignment.append((i, j))
+    alignment.reverse()
```
You can combine this line with the next one by doing this:
```python
return reversed(alignment)
```

okz12 replied:
Made the change and added list() around it to return a list rather than an iterator.

iliakur commented on nltk/metrics/distance.py:
```diff
@@ -103,6 +103,76 @@ def edit_distance(s1, s2, substitution_cost=1, transpositions=False):
+    # set up a 2-D array
+    len1 = len(s1)
+    len2 = len(s2)
+    lev = _edit_dist_init(len1 + 1, len2 + 1)
```
To me the role of this variable would be clearer if it were called `lev_map`, just like you call it in the backtrace function.

okz12 replied:
Good spot. I changed lev_map to lev in the backtrace function because the other edit distance functions (init / step etc.) all just use lev as the variable name. Backtrace was the odd one out.

iliakur commented:
@okz12 thanks, this looks good! I left a few minor comments. It would be cool if you could add some more *strange* cases for the tests. For instance, what happens if one of the strings is empty? It may also be worthwhile to mention in the docstring for the function what happens if multiple alignments are possible for the same minimal edit distance.

okz12 committed:
Changed backtrace to be more declarative

okz12 committed:
Added more test cases for metrics/edit_distance_align

okz12 committed:
Added operator import

okz12 committed:
Corrected test case for new order of precedence in backtrace (skip before substitution)

okz12 commented:
Hey, thanks for the review, that was very helpful. I've incorporated the 3 code comments, added some more test cases and tried to explain the order of backtrace precedence in docstrings for multiple alignments (still not sure if it's clear enough, might need to think of a good example to bring it to life).

stevenbird commented:
Thanks @okz12 and @Copper-Head, this looks good now.
